### PR TITLE
Partial rewrite of fw_sync_with_auth

### PR DESCRIPTION
**Pull request by acv**
This code needs a lot of testing and probably has remaining bugs but it attempts to avoid being affected by client_list changes in any way.

The trick is that the entire client list gets duplicated and the mutex is dropped and then the copy is processed.

Mutex are re-acquired and the client refound (by mac and ip) before being changed. Explicit checks exist for client that no longer exist.

This aims at resolving #84

### CODE REVIEW

**Comment by mhaas** [comment]
Damn, you were about 10 seconds faster than me. Just finished my compile :)

**Comment by acv** [comment]
Last few commits mostly deal with issue #55 at least in theory

**Comment by florida63** [comment]
I test

**Comment by florida63** [comment]
@acv: It seems avoided duplicates rules when I do multiple connections disconnections. cool :)
However, a logout Takes effect only when the gateway receives counter update of server.

Currently no bug.

File: src/http.c [hunk 1]
```diff
@@ -269,7 +269,7 @@ http_callback_auth(httpd * webserver, request * r)
 
             if ((client = client_list_find(r->clientAddr, mac)) == NULL) {
                 debug(LOG_DEBUG, "New client for %s", r->clientAddr);
-                client_list_append(r->clientAddr, mac, token->value);
+                client_list_add(r->clientAddr, mac, token->value);
             } else if (logout) {
                 t_authresponse authresponse;
```
Comment (mhaas) [inline src/http.c]: Something I have noticed here: can't we use logout_client from firewall.c here?

**Comment by acv** [comment]
Indeed! Bonus point, the code in firewall.c was cleaner and with better thread safety.

**Comment by mhaas** [comment]
Ok, I am all happy now that I have #173.

**Pull request merged**